flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateDownloader.java

```diff
@@ -90,27 +80,62 @@ private void downloadDataForAllStateHandles(
             } else {
                 throw new FlinkRuntimeException("Failed to download data for state handles.", e);
             }
+        } finally {
+            // Unregister and close the internal closer. In a failure case, this should interrupt
+            // ongoing downloads.
+            closeableRegistry.unregisterCloseable(internalCloser);
+            IOUtils.closeQuietly(internalCloser);
+            if (failureCleanupRequired) {
+                // Cleanup on exception: cancel all tasks and delete the created directories
+                futures.forEach(future -> future.cancel(true));
+                downloadRequests.stream()
+                        .map(StateHandleDownloadSpec::getDownloadDestination)
+                        .map(Path::toFile)
+                        .forEach(FileUtils::deleteDirectoryQuietly);
+            }
         }
     }
 
-    private List<Runnable> createDownloadRunnables(
-            Map<StateHandleID, StreamStateHandle> stateHandleMap,
-            Path restoreInstancePath,
+    /** Asynchronously runs the specified download requests on executorService. */
+    private Stream<CompletableFuture<Void>> transferAllStateDataToDirectoryAsync(
+            Collection<StateHandleDownloadSpec> handleWithPaths,
             CloseableRegistry closeableRegistry) {
-        List<Runnable> runnables = new ArrayList<>(stateHandleMap.size());
-        for (Map.Entry<StateHandleID, StreamStateHandle> entry : stateHandleMap.entrySet()) {
-            StateHandleID stateHandleID = entry.getKey();
-            StreamStateHandle remoteFileHandle = entry.getValue();
-
-            Path path = restoreInstancePath.resolve(stateHandleID.toString());
+        return createDownloadRunnables(handleWithPaths, closeableRegistry)
+                .map(runnable -> CompletableFuture.runAsync(runnable, executorService));
```

Comment:
Should we now increase `state.backend.rocksdb.checkpoint.transfer.thread.num` ?
Or maybe use `ThreadPoolExecutor` with a bigger maximum size in `RocksDBStateDataTransfer`?

The default is `4`, which limits the benefits that this PR brings.

Reply:
It's hard to say what the right number should be and it's probably different on different systems. Since it's a small change I'd keep it separate from this PR.